CI: rfl: add more tools and steps

ojeda opened the pull request:
This will add some time for the tool building -- the actual steps should be quick, though, and allows us to cover quite a few more tools and unstable features in use.

Please see the individual commits for a few details.

Cc: @GuillaumeGomez @tgross35
r? @Kobzol
try-job: x86_64-rust-for-linux

ojeda committed:
CI: rfl: add a couple comments to split the steps and document them

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

ojeda committed:
CI: rfl: add a Clippy build

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

ojeda committed:
CI: rfl: add a `rustfmt` run

This change will also remove the current warnings in the build due to
`rustfmt` not being available (for `bindgen` output):

    error: 'rustfmt' is not installed for the custom toolchain 'local'.
    note: this is a custom toolchain, which cannot use `rustup component add`
    help: if you built this toolchain from source, and used `rustup toolchain link`, then you may be able to build the component with `x.py`
    Failed to run rustfmt: Internal rustfmt error (non-fatal, continuing)

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

ojeda committed:
CI: rfl: add macro expanded source build (`-Zunpretty=expanded`)

This particular target does not expand into much code, so it is a good
first candidate to see if we could keep this in the CI.

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

ojeda committed:
CI: rfl: factor out build targets

It will make it easier to add more in the future.

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit 9c011ce883055282c5e2453afb2b4fc61d7c97cc with merge 4f7d0c2ca450da0e8500345cb67b3d286c21686b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"9c011ce883055282c5e2453afb2b4fc61d7c97cc","merge_sha":"4f7d0c2ca450da0e8500345cb67b3d286c21686b"} -->

rust-log-analyzer commented:
The job **`x86_64-rust-for-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10332826678/job/28604496165) [(plain)](https://github.com/rust-lang-ci/rust/commit/4f7d0c2ca450da0e8500345cb67b3d286c21686b/checks/28604496165/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
"

# Build a few Rust targets
#
# This does not include building the C side of the kernel nor linking,
#
#
# This includes transforming `rustdoc` tests into KUnit ones thanks to
# `CONFIG_RUST_KERNEL_DOCTESTS=y` above (which, for the moment, uses the
# unstable `--test-builder` and `--no-run`).
    $BUILD_TARGETS

# Generate documentation
make -C linux LLVM=1 -j$(($(nproc) + 1)) \
make -C linux LLVM=1 -j$(($(nproc) + 1)) \
    rustdoc

# Build macro expanded source (`-Zunpretty=expanded`)
# This target also formats the macro expanded code, thus it is also
# This target also formats the macro expanded code, thus it is also
# intended to catch ICEs with formatting `-Zunpretty=expanded` output
# like https://github.com/rust-lang/rustfmt/issues/6105.
    samples/rust/rust_minimal.rsi

# Re-build with Clippy enabled
#
#
# This should not introduce Clippy errors, since `CONFIG_WERROR` is not
# set (thus no `-Dwarnings`) and the kernel uses `-W` for all Clippy
# lints, including `clippy::all`. However, it could catch ICEs.
make -C linux LLVM=1 -j$(($(nproc) + 1)) CLIPPY=1 \
    $BUILD_TARGETS
# Format the code
#
# This returns successfully even if there were changes, i.e. it is not
# a check.
---
'-3dnow' is not a recognized feature for this target (ignoring feature)
'-3dnowa' is not a recognized feature for this target (ignoring feature)
'-3dnow' is not a recognized feature for this target (ignoring feature)
'-3dnowa' is not a recognized feature for this target (ignoring feature)
/home/user/.rustup/toolchains/local/bin/rustfmt: error while loading shared libraries: librustc_driver-394fda38fbbb425c.so: cannot open shared object file: No such file or directory
Failed to run rustfmt: Internal rustfmt error (non-fatal, continuing)
/home/user/.rustup/toolchains/local/bin/rustfmt: error while loading shared libraries: librustc_driver-394fda38fbbb425c.so: cannot open shared object file: No such file or directory
Failed to run rustfmt: Internal rustfmt error (non-fatal, continuing)
/home/user/.rustup/toolchains/local/bin/rustfmt: error while loading shared libraries: librustc_driver-394fda38fbbb425c.so: cannot open shared object file: No such file or directory
'-3dnow' is not a recognized feature for this target (ignoring feature)
'-3dnowa' is not a recognized feature for this target (ignoring feature)
'-3dnow' is not a recognized feature for this target (ignoring feature)
'-3dnowa' is not a recognized feature for this target (ignoring feature)
---
'-3dnow' is not a recognized feature for this target (ignoring feature)
'-3dnowa' is not a recognized feature for this target (ignoring feature)
'-3dnow' is not a recognized feature for this target (ignoring feature)
'-3dnowa' is not a recognized feature for this target (ignoring feature)
/home/user/.rustup/toolchains/local/bin/rustfmt: error while loading shared libraries: librustc_driver-394fda38fbbb425c.so: cannot open shared object file: No such file or directory
make[4]: *** [scripts/Makefile.build:302: samples/rust/rust_minimal.rsi] Error 127
make[4]: *** Deleting file 'samples/rust/rust_minimal.rsi'
make[3]: *** [scripts/Makefile.build:485: samples/rust] Error 2
make[2]: *** [scripts/Makefile.build:485: samples] Error 2
make[1]: *** [/checkout/obj/rfl/linux/Makefile:1925: .] Error 2
make: *** [Makefile:224: __sub-make] Error 2
  local time: Sat Aug 10 16:32:01 UTC 2024
  network time: Sat, 10 Aug 2024 16:32:01 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10332826678/job/28604846330)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10332826678/job/28604846330","builder_name":"checks-actions"} -->

ojeda committed:
CI: rfl: switch to a stage 2 build

Apparently tools like `rustfmt` require it in order to find the right
`librustc_driver.so` without extra tweaks.

Signed-off-by: Miguel Ojeda <ojeda@kernel.org>

ojeda commented:
@bors try

bors commented:
:hourglass: Trying commit 778ff06ba6fe5d5c85604e569179100e00dd24d1 with merge 7ec48b5eef0f322161f7e9e7bbf524b1073f09ec...
<!-- homu: {"type":"TryBuildStarted","head_sha":"778ff06ba6fe5d5c85604e569179100e00dd24d1","merge_sha":"7ec48b5eef0f322161f7e9e7bbf524b1073f09ec"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10590902383/job/29348347143)
Build commit: 7ec48b5eef0f322161f7e9e7bbf524b1073f09ec (`7ec48b5eef0f322161f7e9e7bbf524b1073f09ec`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10590902383/job/29348347143"},"merge_sha":"7ec48b5eef0f322161f7e9e7bbf524b1073f09ec"} -->

ojeda commented:
@Kobzol Moving to `--stage 2` would work, but I am not sure if it is an acceptable cost -- it seems to increase from ~17m to ~30m.

Kobzol commented:
Ok fixing this "bug" (it's probably intended at the moment) would amount to rewriting half of bootstrap, apparently. So let's just use stage2 for now.

@bors r+

bors commented:
:pushpin: Commit 778ff06ba6fe5d5c85604e569179100e00dd24d1 has been approved by `Kobzol`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Kobzol 778ff06ba6fe5d5c85604e569179100e00dd24d1 -->
<!-- homu: {"type":"Approved","sha":"778ff06ba6fe5d5c85604e569179100e00dd24d1","approver":"Kobzol","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
